test/NuGet.Tests.Apex/NuGet.Tests.Apex/NuGetIVsApiTests/NuGetPackageSigningTestCase.cs

```diff
@@ -0,0 +1,412 @@
+                {
+                    var packageToSign = Utils.CreatePackage(packageName, packageVersion1);
+                    await SignedArchiveTestUtility.CreateSignedPackageAsync(testCertificate, packageToSign, pathContext.PackageSource, packageToSign.PackageName);
+                }).GetAwaiter().GetResult();
+
+                Utils.CreatePackageInSource(pathContext.PackageSource, packageName, packageVersion2);
+
+                var nugetTestService = GetNuGetTestService();
+                Assert.True(nugetTestService.EnsurePackageManagerConsoleIsOpen());
+
+                var nugetConsole = nugetTestService.GetPackageManagerConsole(project.UniqueName);
+
+                Assert.True(nugetConsole.InstallPackageFromPMC(packageName, packageVersion2));
+                project.Build();
+                Assert.True(Utils.IsPackageInstalled(nugetConsole, project.FullPath, packageName, packageVersion2));
+
+                Assert.True(nugetConsole.UpdatePackageFromPMC(packageName, packageVersion1));
+                project.Build();
+
+                Assert.False(Utils.IsPackageInstalled(nugetConsole, project.FullPath, packageName, packageVersion2));
+                Assert.True(Utils.IsPackageInstalled(nugetConsole, project.FullPath, packageName, packageVersion1));
+
+                nugetConsole.Clear();
+                solutionService.Save();
+            }
+        }
+
+        [CIOnlyNuGetWpfTheory]
+        [InlineData(ProjectTemplate.ClassLibrary)]
+        [InlineData(ProjectTemplate.NetCoreConsoleApp)]
+        [InlineData(ProjectTemplate.NetStandardClassLib)]
+        public void InstallSignedPackageWithExpiredCertificateFromPMC(ProjectTemplate projectTemplate)
+        {
+            var packageName = "TestPackage";
+            var packageVersion = "1.0.0";
+
+            using (var pathContext = new SimpleTestPathContext())
+            {
+                // Arrange
+                EnsureVisualStudioHost();
+                var solutionService = VisualStudio.Get<SolutionService>();
+
+                solutionService.CreateEmptySolution("TestSolution", pathContext.SolutionRoot);
+                var project = solutionService.AddProject(ProjectLanguage.CSharp, projectTemplate, ProjectTargetFramework.V46, "TestProject");
+                project.Build();
+
+                Task.Run(async () =>
+                {
+                    var packageToSign = Utils.CreatePackage(packageName, packageVersion);
+                    var certWillExpire = SigningTestUtility.GenerateTrustedTestCertificateThatExpiresIn15Seconds();
+                    using (var testCertificate = new X509Certificate2(certWillExpire.Source.Cert))
+                    {
+                        await SignedArchiveTestUtility.CreateSignedPackageAsync(testCertificate, packageToSign, pathContext.PackageSource, packageToSign.PackageName);
+                    }
+
+                    // Wait for cert to expire
+                    Thread.Sleep(15000);
```

Comment:
This is a long wait for something that is tested for 3 templates. I would either make this for 1 template, or create just 1 expired package and re-use it for all tests.

Reply:
I think it would be easier and more robust to create a package with a certificate that will expire immediately after signing and then add that signed package as a static test resource.  If there's a problem with trust you could extract the certificate chain and temporarily add the root as a trusted root authority.